flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java

```diff
@@ -67,361 +89,422 @@
-				}
-
-				if (LOG.isDebugEnabled()) {
-					LOG.debug("{} (taskIdx={}) restored {}.", getClass().getSimpleName(), subtaskIdx, restoredReaderState);
-				}
-			}
-		} else {
+		if (!context.isRestored()) {
 			LOG.info("No state to restore for the {} (taskIdx={}).", getClass().getSimpleName(), subtaskIdx);
+			return;
+		}
+
+		LOG.info("Restoring state for the {} (taskIdx={}).", getClass().getSimpleName(), subtaskIdx);
+
+		splits = splits == null ? new PriorityQueue<>() : splits;
+		for (TimestampedFileInputSplit split : checkpointedState.get()) {
+			splits.add(split);
+		}
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("{} (taskIdx={}) restored {}.", getClass().getSimpleName(), subtaskIdx, splits);
 		}
 	}
 
 	@Override
 	public void open() throws Exception {
 		super.open();
 
-		checkState(this.reader == null, "The reader is already initialized.");
 		checkState(this.serializer != null, "The serializer has not been set. " +
 			"Probably the setOutputType() was not called. Please report it.");
 
+		this.state = ReaderState.IDLE;
 		this.format.setRuntimeContext(getRuntimeContext());
 		this.format.configure(new Configuration());
-		this.checkpointLock = getContainingTask().getCheckpointLock();
-
-		// set the reader context based on the time characteristic
-		final TimeCharacteristic timeCharacteristic = getOperatorConfig().getTimeCharacteristic();
-		final long watermarkInterval = getRuntimeContext().getExecutionConfig().getAutoWatermarkInterval();
-		this.readerContext = StreamSourceContexts.getSourceContext(
-			timeCharacteristic,
-			getProcessingTimeService(),
-			checkpointLock,
-			getContainingTask().getStreamStatusMaintainer(),
-			output,
-			watermarkInterval,
-			-1);
-
-		// and initialize the split reading thread
-		this.reader = new SplitReader<>(format, serializer, readerContext, checkpointLock, restoredReaderState);
-		this.restoredReaderState = null;
-		this.reader.start();
+
+		this.sourceContext = StreamSourceContexts.getSourceContext(
+				getOperatorConfig().getTimeCharacteristic(),
+				getProcessingTimeService(),
+				new Object(), // no actual locking needed
+				getContainingTask().getStreamStatusMaintainer(),
+				output,
+				getRuntimeContext().getExecutionConfig().getAutoWatermarkInterval(),
+				-1);
+
+		this.reusedRecord = serializer.createInstance();
+		this.completedSplitsCounter = getMetricGroup().counter("numSplitsProcessed");
+		this.executor = executor != null ? executor : getContainingTask().getMailboxExecutorFactory().createExecutor(getOperatorConfig().getChainIndex());
+		this.splits = this.splits == null ? new PriorityQueue<>() : this.splits;
+
+		if (!splits.isEmpty()) {
+			enqueueMail();
+		}
 	}
 
 	@Override
 	public void processElement(StreamRecord<TimestampedFileInputSplit> element) throws Exception {
-		reader.addSplit(element.getValue());
+		Preconditions.checkState(state.isAcceptingSplits());
+		splits.offer(element.getValue());
+		if (state == ReaderState.IDLE) {
+			enqueueMail();
```

Comment:
this call happens already in the mailbox thread. Is there a reason why it has to be enqueued to the mailbox queue?

Reply:
For each record in a file split, a mail has to be enqueued to allow process other mails.
After receiving a split we **could** process its 1st record right away. We'd need to check if there are other splits being processed and if this split is empty. 
But the potential speedup is negligible (just 1st record), so I'd rather avoid even a single extra if.